egi_notebooks_hub/d4science.py

```diff
@@ -151,3 +152,193 @@ def pre_spawn_start(self, user, spawner):
+        for jwk in jwks_keys:
+            kid = jwk["kid"]
+            self._pubkeys[kid] = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk))
+        return self._pubkeys
+
+    async def authenticate(self, handler, data=None):
+        # first get authorized upstream
+        user_data = await super().authenticate(handler, data)
+        context = getattr(self, "d4science_context", None)
+        if not context:
+            self.log.error("Unable to get the user context")
+            raise web.HTTPError(403)
+        self.log.debug("Context is %s", context)
+        # TODO: do we need to check anything on the context?
+        body = urlencode(
+            {
+                "grant_type": "urn:ietf:params:oauth:grant-type:uma-ticket",
+                "claim_token_format": "urn:ietf:params:oauth:token-type:jwt",
+                "audience": self.client_id,
+                "claim_token": base64.b64encode(
+                    json.dumps({"context": [f"{context}"]}).encode("utf-8")
+                ),
+            }
+        )
+        http_client = AsyncHTTPClient()
+        req = HTTPRequest(
+            self.token_url,
+            method="POST",
+            auth_username=self.client_id,
+            auth_password=self.client_secret,
+            auth_mode="basic",
+            headers={
+                "Content-Type": "application/x-www-form-urlencoded; charset=utf-8"
+            },
+            body=body,
+        )
+        try:
+            resp = await http_client.fetch(req)
+        except HTTPError as e:
+            # whatever, get out
+            self.log.warning("Unable to get the permission for user: %s", e)
+            raise web.HTTPError(403)
+        self.log.debug("Got UMA ticket from server...")
+        token = json.loads(resp.body.decode("utf8", "replace"))["access_token"]
+        kid = jwt.get_unverified_header(token)["kid"]
+        key = (await self.get_iam_public_keys())[kid]
+        decoded_token = jwt.decode(
+            token,
+            key=key,
+            audience=self.client_id,
+            algorithms=["RS256"],
+        )
+        # TODO: add extra checks?
+        permissions = decoded_token["authorization"]["permissions"]
+        user_data["auth_state"].update(
+            {
+                "uma_token": token,
+                "permissions": permissions,
+                "context": context,
+            }
+        )
+        return user_data
+
+    async def pre_spawn_start(self, user, spawner):
+        """Pass gcube-token to spawner via environment variable"""
+        auth_state = await user.get_auth_state()
+        if not auth_state:
+            # auth_state not enabled
+            return
+        spawner.environment["GCUBE_TOKEN"] = auth_state["uma_token"]
+        # spawner.environment["DATAMINER_URL"] = auth_state["wps-endpoint"]
+        spawner.environment["GCUBE_VRE"] = unquote(auth_state["context"])
+
+
+class D4ScienceSpawner(KubeSpawner):
+    frame_ancestors = Unicode(
```

Comment:
can we also add https://www.google-analytics.com here? it has been reported a problem with google analytics and CORS https://support.d4science.org/issues/21502

Reply:
As default? Is google analytics a frame ancestor?

Reply:
ok no worries, it's just google analytics complaining on the js console. no arm